### PR TITLE
ACME: new helper module for ACME challenges which need TLS certs

### PR DESCRIPTION
**Pull request by felixfontein**
##### SUMMARY
Continuation of #42158. Adds a new module which allows to create the required challenge TLS certs needed for the [ACME TLS-ALPN-01 challenge](https://tools.ietf.org/html/draft-ietf-acme-tls-alpn-01). ([Current draft 05](https://tools.ietf.org/html/draft-ietf-acme-tls-alpn-05)) 

##### ISSUE TYPE
 - New Module Pull Request

##### COMPONENT NAME
acme_challenge_cert_helper

##### ANSIBLE VERSION
```
2.7.0
```

### CODE REVIEW

**Comment by felixfontein** [comment]
I'll add tests once ansible/acme-test-container#5 is merged.

**Comment by felixfontein** [comment]
My current (local) test environment for this module is this part uncommented (and the one below commented out): https://github.com/felixfontein/ansible-acme-test/blob/master/test-certificate-obtain.yml#L86

**Comment by felixfontein** [comment]
CC @resmo

**Review by resmo — APPROVED** [review]
shipit

File: lib/ansible/modules/web_infrastructure/acme_challenge_cert_helper.py
Comment (resmo) [inline lib/ansible/modules/web_infrastructure/acme_challenge_cert_helper.py]: leftover?

File: lib/ansible/modules/web_infrastructure/acme_challenge_cert_helper.py
Comment (resmo) [inline lib/ansible/modules/web_infrastructure/acme_challenge_cert_helper.py]: minor whitespace

Comment (felixfontein) [reply lib/ansible/modules/web_infrastructure/acme_challenge_cert_helper.py]: Thanks, fixed it.

File: lib/ansible/modules/web_infrastructure/acme_challenge_cert_helper.py
Comment (felixfontein) [reply lib/ansible/modules/web_infrastructure/acme_challenge_cert_helper.py]: No, not really, it is more a mark for "insert whatever module you need to do this here" ;-) (might require more than one module) How this task would look like depends a lot on the webserver of choice and how it is configured. I want to try this out with NGINX at some point to write a how-to on how this can actually be done, but so far I didn't got around to do that :)

**Comment by felixfontein** [comment]
@resmo The module isn't ready to be published yet, because the Object ID for the TLS extension isn't reserved yet. A reviewer of the draft-ietf-acme-tls-alpn-01 draft noticed some days ago that the OID (1.3.6.1.5.5.7.1.30) mentioned in the draft is already used for something else ([link to registry](https://www.iana.org/assignments/smi-numbers/smi-numbers.xml#smi-numbers-1.3.6.1.5.5.7.1)); there's a new draft, but now another expert mentioned that the OID in there isn't reserved for this draft either, so it might change again. I'd like to wait until that is settled (and the final OID has been propagated to [Pebble](https://github.com/letsencrypt/Pebble) and [Boulder](https://github.com/letsencrypt/boulder/), and into the ACME test container) before shipping this module.

**Comment by resmo** [comment]
@felixfontein thanks for the info

(WIP needs to be prefix-ish to be recognised by the bot IMHO)

**Comment by felixfontein** [comment]
(Forgot a reference to the discussion: https://mailarchive.ietf.org/arch/msg/acme/--P1aNB-1pTe-sgLjR7nVQKvlxM)

@resmo Also, I was wondering whether this module should be in `web_infrastructure` (as it is right now), or in `crypto`. The main plus factor of `web_infrastructure` is that all other ACME modules are in there, and while this module usually fits better into `crypto` (because it does stuff similar to other modules in there), it has no real use without `acme_certificate`. I'd prefer it to be with the other `acme_*` modules, but I wanted to raise this point to see what you think about it.

(And thanks for moving the [WIP] to the correct place :) )

**Comment by felixfontein** [comment]
(The good thing about the discussions on the ML is that it doesn't affect the TLS-ALPN-01 part in `acme_certificate` itself. :) )

**Comment by resmo** [comment]
I also wondered if it would make sense to move all acme_ to /crypto (this path didn't exist at that time) since acme_ is not (only) related to web. I would be ok to move all to /crypto. It has no impact to the users.

**Comment by felixfontein** [comment]
Good question :) I always thought it was in `web_infrastructure` for a good reason ;)

@Spredzy, since you're the maintainer for `crypto`: what do you think about this?

**Comment by felixfontein** [comment]
Ah, so a plain prefix does the trick :)

**Comment by Spredzy** [comment]
@felixfontein it totally does make sense. Feel free to do the PR and add yourself in the maintainer group.

**Comment by felixfontein** [comment]
@Spredzy @resmo I created a PR in #44279. I also moved this module to `crypto`.

**Comment by felixfontein** [comment]
Ooops, there was something wrong with the last commit... Should be fixed now.

**Comment by felixfontein** [comment]
@resmo I think it's done now. The [OID has been officially assigned](https://www.iana.org/assignments/smi-numbers/smi-numbers.xhtml#smi-numbers-1.3.6.1.5.5.7.1), so it shouldn't change anymore in the future. Also, Pebble supports both OIDs now, and support for Boulder has been merged (letsencrypt/boulder#3819). It will surely be deployed before Ansible 2.7 is released.

**Review by resmo — APPROVED** [review]
shipit

**Comment by felixfontein** [comment]
@resmo Thanks for reviewing and merging!

**Pull request merged**
